### PR TITLE
New module os_group_info

### PR DESCRIPTION
**Pull request by phsmith**
##### SUMMARY
Added new module named os_group_facts, based on os_project_facts, to retrieve groups data.
Fixed os_user_role module problem that didn't validate groups correctly.

##### ISSUE TYPE
- Bugfix Pull Request
- New Module Pull Request

##### COMPONENT NAME
os_project_facts

### CODE REVIEW

File: lib/ansible/modules/cloud/openstack/os_group_facts.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+          this dictionary may be additional dictionaries.
+   availability_zone:
+     description:
+       - Ignored. Present for backwards compatibility
+'''
+
+EXAMPLES = '''
+# Gather facts about previously created groups
+- os_group_facts:
+    cloud: awesomecloud
```
Comment (cloudnull) [inline lib/ansible/modules/cloud/openstack/os_group_facts.py]: do these example tasks need a "register"?

Comment (phsmith) [reply lib/ansible/modules/cloud/openstack/os_group_facts.py]: @cloudnull Sorry for the late answer.
No, this module don't need a register, It returns a variable named "openstack_groups".

**Review by StingRayZA — COMMENTED** [review]
LGTM, and would help me out...

**Comment by StingRayZA** [comment]
@phsmith Can you close and reopen this PR to get the CI to freshen up please? once done, I'd like to look for a maintainer to shipit...

**Comment by phsmith** [comment]
Done @StingRayZA
Thanks for the attention.

File: lib/ansible/modules/cloud/openstack/os_group_facts.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+description:
+    - Retrieve facts about a one or more OpenStack groups
+requirements:
+    - "python >= 2.7"
+    - "openstacksdk"
+options:
+   name:
+     description:
+        - Name or ID of the group
+     required: true
```
Comment (StingRayZA) [inline lib/ansible/modules/cloud/openstack/os_group_facts.py]:
```suggestion
     required: true
     type: str
```

File: lib/ansible/modules/cloud/openstack/os_group_facts.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+   domain:
+     description:
+        - Name or ID of the domain containing the group if the cloud supports domains
```
Comment (StingRayZA) [inline lib/ansible/modules/cloud/openstack/os_group_facts.py]:
```suggestion
        - Name or ID of the domain containing the group if the cloud supports domains
     type: str
```

File: lib/ansible/modules/cloud/openstack/os_group_facts.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+   filters:
+     description:
+        - A dictionary of meta data to use for further filtering.  Elements of
+          this dictionary may be additional dictionaries.
```
Comment (StingRayZA) [inline lib/ansible/modules/cloud/openstack/os_group_facts.py]:
```suggestion
          this dictionary may be additional dictionaries.
     type: dict
```

File: lib/ansible/modules/cloud/openstack/os_group_facts.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+   availability_zone:
+     description:
+       - Ignored. Present for backwards compatibility
```
Comment (StingRayZA) [inline lib/ansible/modules/cloud/openstack/os_group_facts.py]:
```suggestion
       - Ignored. Present for backwards compatibility
     type: str
```

**Review by StingRayZA — COMMENTED** [review]
Hopefully these fixes should satisfy the linter :)

**Review by StingRayZA — APPROVED** [review]
LGTM. Let's find a couple of maintainers to shipit. 😁

**Comment by phsmith** [comment]
Nice @StingRayZA! Thanks!

**Comment by StingRayZA** [comment]
Hi @phsmith 
I've been chatting to @felixfontein on IRC:
```
<felixfontein> spontaneously: that should be a _info module, not a _facts module
<stingrayza> felixfontein: thanks. If I could ask why? (the rest of the openstack modules that are pure 'retrieve info' mods are _facts ) https://docs.ansible.com/ansible/latest/modules/list_of_cloud_modules.html#openstack
<felixfontein> the rest will be renamed eventually before 2.9 is released
<felixfontein> I think they all return ansible_facts, that's why we're waiting for a new feature before renaming them
<felixfontein> the reason is that _facts should only be used for modules which return facts about the system they run on. cloud modules return information on remote things, which do not depend on the host the module is executed on
<stingrayza> so, instead of returning _facts, we should return a value which could be registered?
<felixfontein> yes
<stingrayza> great feedback, thanks! (mind if I quote you on the PR? - it's not my PR, but I want the module ;) )
<felixfontein> sure, feel free to CC me :)
```
There's this Issue for reference - #54280

As a result of that conversation, I recommend the following:

* Change the module name to os_group_info
* Change the `module.exit_json` (line 147-148) as below:

from 
```
         module.exit_json(changed=False, ansible_facts=dict(
            openstack_groups=groups))
```
to
```
         module.exit_json(changed=False, groups=groups)
```

* Change the return values (from line 81)

to
```
RETURN = '''
groups:
    description: Dictionary describing all the matching groups.
    returned: always, but can be null
    type: complex
    contains:
```

* Change the examples (from line 47) to include a register, e.g.

```
 # Gather facts about previously created groups
 - os_group_facts:
     cloud: awesomecloud
   register: openstack_groups
 - debug:
     var: openstack_groups
```

* Last item I just noticed is (line 2) - your Copyright statement. Unless you're working for HP, that should probably read something like the following (see os_volume_snapshot.py as an example)

```
# Copyright (c) 2019, Phillipe Smith <phillipelnx@gmail.com>
```

Hope this all makes sense, feel free to ping me with any questions.
This should, hopefully, be the last change 🤞 😊

**Review by StingRayZA — APPROVED** [review]
great, let's see what the maintainers have to say :)

**Comment by phsmith** [comment]
Thanks a lot for the support @StingRayZA

**Review by emonty — APPROVED** [review]
shipit

**Comment by gtema** [comment]
shipit

**Comment by StingRayZA** [comment]
:shipit: 😁

File: lib/ansible/modules/cloud/openstack/os_group_info.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        - Name or ID of the domain containing the group if the cloud supports domains
+     type: str
+   filters:
+     description:
+        - A dictionary of meta data to use for further filtering.  Elements of
+          this dictionary may be additional dictionaries.
+     type: dict
+   availability_zone:
+     description:
+       - Ignored. Present for backwards compatibility
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/openstack/os_group_info.py:44]: Does it make sense to have this in a new module?

File: lib/ansible/modules/cloud/openstack/os_group_info.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    description: Dictionary describing all the matching groups.
+    returned: always, but can be null
+    type: complex
+    contains:
+        name:
+            description: Name given to the group.
+            returned: success
+            type: string
+        description:
+            description: Description of the group
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/openstack/os_group_info.py]:
```suggestion
            description: Description of the group.
```
You added a period most of the time, but not here (and for `domain_id`'s description).

File: lib/ansible/modules/cloud/openstack/os_user_role.py [hunk 8]
```diff
@@ -151,7 +151,7 @@ def main():
                 module.fail_json(msg="User %s is not valid" % user)
             filters['user'] = u['id']
         if group:
-            g = cloud.get_group(group)
+            g = cloud.get_group(group, domain_id=filters['domain'])
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/openstack/os_user_role.py]: In which way is this a fix? You should add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to) for this. And maybe it's better to do this in a separate PR, so this fix can be backported.

File: lib/ansible/modules/cloud/openstack/os_group_info.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+DOCUMENTATION = '''
+---
+module: os_group_info
+short_description: Retrieve info about one or more OpenStack groups
+extends_documentation_fragment: openstack
+version_added: "2.9"
+author: "Phillipe Smith (@phsmith)"
+description:
+    - Retrieve info about a one or more OpenStack groups
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/openstack/os_group_info.py]:
```suggestion
    - Retrieve info about a one or more OpenStack groups.
```
The description should have full sentences.

File: lib/ansible/modules/cloud/openstack/os_group_info.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+requirements:
+    - "python >= 2.7"
+    - "openstacksdk"
+options:
+   name:
+     description:
+        - Name or ID of the group
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/openstack/os_group_info.py]:
```suggestion
        - Name or ID of the group.
```
(and some more)

**Review by felixfontein — COMMENTED** [review]
LGTM in general. And it's much better than an `_info` module :-)

File: lib/ansible/modules/cloud/openstack/os_group_info.py (hunk 6 above)
Comment (phsmith) [reply lib/ansible/modules/cloud/openstack/os_group_info.py:44]: Don't thing so. Removed that option.

**Comment by felixfontein** [comment]
`openstack_full_argument_spec` seems to include this parameter; the sanity tests now complain that it isn't documented. Shouldn't this be a part of the `openstack` documentation fragment, then? Anyway, I guess you can re-add that parameter (i.e. revert 20155c0b0f5b5b7ec88a9ef4abe1fb56a56ccd87) for now.

OpenStack maintainers: if this argument isn't needed, wouldn't it be a good idea to deprecate it in modules which already have it, and avoid adding it to new modules?

File: lib/ansible/modules/cloud/openstack/os_user_role.py (hunk 8 above)
Comment (phsmith) [reply lib/ansible/modules/cloud/openstack/os_user_role.py]: You're right @felixfontein. 
Reverted the code to open a new PR.

**Comment by felixfontein** [comment]
@emoty @gtema @StingRayZA is this still ready to go?

**Comment by StingRayZA** [comment]
looking good from my end... just needs a shipit from a maintainer or few 😀

**Comment by emonty** [comment]
shipit

**Comment by gtema** [comment]
ship

**Comment by felixfontein** [comment]
"ship" won't work, you need to write "shipit". But since bot won't automerge anyway, I'll just assume you mean "shipit" and just merge :)

**Comment by felixfontein** [comment]
@phsmith thanks for contributing this new module!
Everyone else, thanks for reviewing this!

And BTW, it would be nice if someone could eventually clean up `availability_zone`, or at least create an issue for it so progress can be tracked :)

**Comment by gtema** [comment]
> "ship" won't work, you need to write "shipit". But since bot won't automerge anyway, I'll just assume you mean "shipit" and just merge :)

Yeah, I know. crappy auto-correction. Thanks

**Comment by phsmith** [comment]
Thanks guys! Hope that module could be useful.
